### PR TITLE
Load balancers

### PR DESCRIPTION
**Pull request by Cocotton**
The PR has been tested :)

### CODE REVIEW

File: service/arm_templates/main.json [hunk 1]
```diff
@@ -152,9 +162,9 @@
         },
         "parameters": {
           "clusterID": {"value": "[parameters('clusterID')]"},
-          "apiLoadBalancerCidr": {"value": "[parameters('apiLoadBalancerCidr')]"},
+          "apiLoadBalancerCidr": {"value": "[reference('api_load_balancer_setup').outputs.loadBalancerIPAddress.value]"},
           "etcdLoadBalancerCidr": {"value": "[parameters('etcdLoadBalancerCidr')]"},
```
Comment (kopiczko) [inline service/arm_templates/main.json]: Even though it's private, can the cidr be taken from the output for consistency reasons?

File: service/arm_templates/public_load_balancer_setup.json [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      }
+    },
+    "loadBalancerIDOutput": {
+      "type": "string",
+      "value": "[reference(variables('loadBalancerID'), parameters('loadBalancersAPIVersion'), 'Full')]"
+    },
+    "PublicIPIDOutput": {
+      "type": "string",
+      "value": "[reference(variables('loadBalancerPublicIPId'), parameters('publicIPAddressesAPIVersion'), 'Full')]"
+    }
```
Comment (kopiczko) [inline service/arm_templates/public_load_balancer_setup.json]: Do we need those two last outputs?

File: service/arm_templates/public_load_balancer_setup.json [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    {
+      "apiVersion": "[parameters('publicIPAddressesAPIVersion')]",
+      "type": "Microsoft.Network/publicIPAddresses",
+      "name": "[variables('loadBalancerPublicIPName')]",
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "publicIPAllocationMethod": "Dynamic",
+        "publicIPAddressVersion": "IPv4"
+      }
+    },
```
Comment (kopiczko) [inline service/arm_templates/public_load_balancer_setup.json]: How is the IP assigned to the LB?

Comment (Cocotton) [reply service/arm_templates/public_load_balancer_setup.json]: In the "Microsoft.Network/loadBalancers" block, under properties.frontendIPConfigurations[0].properties.publicIPAddress.id

File: service/arm_templates/public_load_balancer_setup.json (hunk 2 above)
Comment (Cocotton) [reply service/arm_templates/public_load_balancer_setup.json]: They were for testing purposes, I'm done with the testing and removed them :)

File: service/arm_templates/private_load_balancer_setup.json [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+          {
+            "name": "[variables('loadBalancerBackendPoolName')]"
+          }
+        ],
+        "copy": [
+          {
+            "name": "loadBalancingRules",
+            "count": "[length(parameters('ports'))]",
+            "input": {
+              "name": "[concat('loadBalancingRule-', string(add(copyIndex('loadBalancingRules'), 1)))]",
```
[the remark this answers is not in the document]
Comment (Cocotton) [reply service/arm_templates/private_load_balancer_setup.json]: Makes sense to me too! Updated it

File: service/arm_templates/main.json [hunk 5]
```diff
@@ -257,6 +270,83 @@
...
+      "type": "Microsoft.Resources/deployments",
+      "properties": {
+        "mode": "incremental",
+        "templateLink": {
+          "uri": "[variables('publicLoadBalancerTemplateURI')]",
+          "contentVersion": "1.0.0.0"
+        },
+        "parameters": {
+          "clusterID": {"value": "[parameters('clusterID')]"},
+          "prefix": {"value": "Ingress-HTTP"},
```
[the remark this answers is not in the document]
Comment (Cocotton) [reply service/arm_templates/main.json]: Fixed

File: service/arm_templates/main.json (hunk 1 above)
Comment (Cocotton) [reply service/arm_templates/main.json]: The logical place from where to take the cidr would be from the network_setup and then pass it to the security_groups_setup template, but that would put us in a deadlock :(

File: service/arm_templates/virtual_network_setup.json [hunk 6]
```diff
@@ -86,6 +121,18 @@
...
+    },
+    "etcdLoadBalancerSubnetID": {
+      "type": "string",
+      "value": "[variables('etcdLoadBalancerSubnetID')]"
+    },
+    "ingressLoadBalancerSubnetID": {
+      "type": "string",
+      "value": "[variables('ingressLoadBalancerSubnetID')]"
+    },
   }
```
Comment (kopiczko) [inline service/arm_templates/virtual_network_setup.json]: I wonder if we need those outputs at all. Do we?

File: service/arm_templates/virtual_network_setup.json [hunk 7]
```diff
@@ -69,7 +86,25 @@
...
+            "properties": {
+              "addressPrefix": "[parameters('etcdLoadBalancerSubnetCidr')]"
+            }
+          },
+          {
+            "name": "[variables('ingressLoadBalancerSubnetName')]",
+            "properties": {
+              "addressPrefix": "[parameters('ingressLoadBalancerSubnetCidr')]"
+            }
+          },
```
Comment (kopiczko) [inline service/arm_templates/virtual_network_setup.json]: Do we need virtual network subnets for ingress and API? They have just static public IP. This is where my network foo ends :P

Comment (Cocotton) [reply service/arm_templates/virtual_network_setup.json]: You are right, I'll remove them! I'll remove them from the azuretpr too, unless you see something useful we could do with them

File: service/arm_templates/virtual_network_setup.json (hunk 6 above)
Comment (Cocotton) [reply service/arm_templates/virtual_network_setup.json]: Will remove that too :)

File: service/arm_templates/main.json [hunk 8]
```diff
@@ -152,9 +156,9 @@
         },
         "parameters": {
           "clusterID": {"value": "[parameters('clusterID')]"},
-          "apiLoadBalancerCidr": {"value": "[parameters('apiLoadBalancerCidr')]"},
+          "apiLoadBalancerCidr": {"value": "[reference('api_load_balancer_setup').outputs.publicIPCidr.value]"},
```
Comment (kopiczko) [inline service/arm_templates/main.json]: I think we don't need this var.

File: service/arm_templates/main.json [hunk 9]
```diff
@@ -152,9 +156,9 @@
...
           "etcdLoadBalancerCidr": {"value": "[parameters('etcdLoadBalancerCidr')]"},
-          "ingressLoadBalancerCidr": {"value": "[parameters('ingressLoadBalancerCidr')]"},
+          "ingressLoadBalancerCidr": {"value": "[reference('ingress_load_balancer_setup').outputs.publicIPCidr.value]"},
```
Comment (kopiczko) [inline service/arm_templates/main.json]: I think we don't need this var.

File: service/arm_templates/public_load_balancer_setup.json [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+      "type": "object",
+      "value": {
+        "id": "[variables('loadBalancerBackendPoolID')]",
+        "name": "[variables('loadBalancerBackendPoolName')]"
+      }
+    },
+    "publicIPCidr": {
+      "type": "string",
+      "value": "[concat(reference(variables('loadBalancerPublicIPId'), parameters('publicIPAddressesAPIVersion')).ipAddress, '/32')]"
+    }
```
Comment (kopiczko) [inline service/arm_templates/public_load_balancer_setup.json]: I think we don't need this output.

**Pull request merged**